### ISSUE TITLE
Sun's Song random notes softlock

### ISSUE THREAD

**Issue by r0bd0g** [TestRunnerSRL/OoT-Randomizer#1792]
I think we got a report of this happening even after our fix for long songs. I might have dreamed it though. I can't find any solid documentation of this happening. Edit: Okay I found it.

**Comment by r0bd0g** [comment]
ETR_BTF — 10/12/2022 9:21 AM
remember that bug with sun's song / random notes where you could not complete certain puzzles and time would not advance properly? flag introduced a fix for that (#1478) which worked for one seed that flag had tested but just now someone reported that bug again. flag's fix seems to be increasing the song's duration by 260 ms(?) but perhaps that's not enough for all variants of suns song? 
here's the log https://discord.com/channels/274180765816848384/274186223147417600/1029729316721930250
it happened on 6.2.133 M-7 (mracsys), long after flag's fix had been merged

Ben Plays — 10/12/2022 9:03 AM
In symphony room in MQ Spirit I can't seem to progress after the sun gate opens. Is there something I'm missing?
I think that room is broken because sun's song isn't working
I have randomized notes
Bliven — 10/12/2022 9:10 AM
There have def been bugs reported with sun song and random notes
Ben Plays — 10/12/2022 9:11 AM
Hmmmm ok
Note to self: Don't do random notes again
Bliven — 10/12/2022 9:11 AM
Is it 8 notes?
Ben Plays — 10/12/2022 9:11 AM
Yes

**Comment by ETR-BTF** [comment]
Duplicate of #1326, which was (in hindsight prematurely) closed when #1478 was merged.

**Comment by r0bd0g** [comment]
Yeah we asked then it be reopened, but nobody did it, so I opened the duplicate when I realized it was being forgotten.

**Comment by fenhl** [comment]
Another instance of this at <https://ootrandomizer.com/seed/get?id=1352991> with a 6-note Sun's. Symphony room didn't work, passing time of day did, not sure about frogs.

**Comment by ETR-BTF** [comment]
Happened again. 7 notes this time. I think it's safe to say the number of notes is irrelevant. However, the only issue was they couldn't trigger the puzzle in MQ Spirit. The song still passed time of day correctly and they could also get the checks in the redead grave and from the river frogs. Maybe flag's fix works for some but not all checks? Is it because it's an MQ dungeon?

https://ootrandomizer.com/seed/get?id=1353129

**Comment by mzxrules** [comment]
I feel fairly confident I've found the true problem. It's pretty complex, spanning multiple systems, but the issue is caused by changing the duration of the sun's song playback. the details are something like this: 

Actor 0x12E (okarina_tag) is the actor responsible for detecting if the player has pulled out the Ocarina/Has played the correct song. In MQ Spirit Temple, the actor swaps between two action function states. 
* `MQD_func_80ABEF2C` detects if the player pulls out the Ocarina in the right spot, and if so calls Message_StartOcarina to pull up the music staff, then switches to the next state.
* `MQD_func_80ABF0CC` checks `play->msgCtx.ocarinaMode` every frame to determine if the message system is done. For the sun's song check, a value of 4 here is the "fail" condition, swapping back to the other state, while a value of 3 is the success state.

Message_DrawMain is responsible for handling the different states of the message box system. The case MSGMODE_SONG_PLAYED_ACT handles returning whether the correct song was played, and setting `play->msgCtx.ocarinaMode` to 3 if so. This is set on the last frame of the textbox, since this case closes the textbox.

Actor 0x17E (z_oceff_spot) is the actor responsible for the cool effect that envelops Link he plays Sun's Song. Importantly, it is responsible for setting `save_context.sunsSongState` to SUNSSONG_START

z_parameter is responsible for a lot of things, but at the end of Interface_Update there is code to handle the Sun's Song time warp effect if if`gSaveContext.sunsSongState` is non-zero. At the start of this block is a check that tests if `save_context.sunsSongState` is SUNSSONG_START, and if so it will set `play->msgCtx.ocarinaMode` to 4

Got all that?

Now the core conflict is between z_oceff_spot and the message system. When you play the correct song, the message system switches to song playback mode. Around the same time, z_oceff_spot is spawned to create the cool light effect. If z_oceff_spot's effect completes before the message system can finish the song playback, `play->msgCtx.ocarinaMode` will be set to 4 first, and okarina_tag will treat it as the wrong song being played.

As for a fix, we have a few options. Easiest and safest would be to clamp playback to the Sun's Song default length of I think 208 frames. The cool option would be to adjust the length of the `OceffSpot_Wait` timer set in `OceffSpot_GrowCylinder` to account for a longer playback; the timer is 60 frames long so you'd have to be careful in how you handle things.

**Comment by r0bd0g** [comment]
This issue still occurs.
Seems likely this time the issue is the song being too short.

bastischo
 — 
Today at 6:25 PM
Hello, I'm new so Ihope I'm not missing any important rules, but I wanted to report a potential bug.
I'm playing with randmoized songs and cannot complete the frog minigames.
I can play all songs successfully except for sun's song which softlocks me.
I think it is because the song has only 5 notes (V,<,A,<,A) so the last frog won't jump and end the cutscene.
This is the seed in case it's relevant: https://ootrandomizer.com/seed/get?id=1813473

bastischo
 — 
Today at 7:38 PM
OMG, Now I'm in MQ SPirit temple and the Ocarina minigame seems to be bugging out because of sun's song as well.

**Comment by cjohnson57** [comment]
Is fixing it as simple as making sun's song a minimum 6 notes?

**Comment by fenhl** [comment]
Probably not. We already increase the duration of the song (time, not notes) if it's below the minimum.

**Comment by fenhl** [comment]
Cannot reproduce using the Sun's Song frog with `v<A<A` as the melody. It might depend on the playback.

**Comment by flagrama** [comment]
@rrealmuto The behaviour seems to have changed from a softlock to a hard crash on enemizer. https://discord.com/channels/274180765816848384/274186223147417600/1446538904118366361

Not entirely sure if it is actually related to this bug report or if it should be a new one.

Screencap of crash screen:

<img width="2002" height="1528" alt="Image" src="https://github.com/user-attachments/assets/613bd18c-cbcb-45f9-9b56-c42f162cef94" />

**Comment by djevangelia** [comment]
I was really lucky and got a bugged seed on first try. The problem is that the (I guess) final note timer length (unsigned) underflows. If you memory edit extra length to compensate for the subtraction, it works fine, and if you change the FFFFFFFF length to zero, frogs work and you get the Sun's Song time of day effect.
https://www.youtube.com/watch?v=e4NqsrkDUqM
It's likely easy to quickly identify an affected seed by playing Sun's Song on spawn and looking at 0x8010223c `sPlaybackNoteTimer` after finishing. I don't have patch file and I don't get how to generate plando from spoiler log.

I'm really astonished that I can't generate a new bugged seed. Is there variability in song length compensation? I'm using the same Sun's Song pattern. I couldn't get it by memory editing length when playing on a normal seed either.

As for a fix, considering how difficult it seems to reproduce a seed both for testing a fix and to find edge cases, maybe clamping the note length somewhere?

**Comment by fenhl** [comment]
Yeah, there are some aspects of note randomization which currently can't be plando'd, see e.g. `OcarinaSongs.random_playback`. Maybe we should include these in spoiler logs and make them plandoable to help debug this.

**Comment by djevangelia** [comment]
Thanks! I enabled the OcarinaSongs print playback data function and was lucky again with two bug seeds with output of the playback data. I inserted this data into the playback_data again and these new clone seeds were also bugged. (I did the same thing with a normal seed as well which didn't turn out bugged.)

So Sun's Song playback was randomized to the Ocarina song datas below (base is AvvAv), underflows, and same behavior when plugged into different seed.

Note: This build has various additions of varying quality. In particular a Wallmaster might spawn approximately every 3 minutes. But it goes away when it returns back up.

OoT_72372_AKDSAIGODV 
[OoT_72372_AKDSAIGODV suns song patch.zip](https://github.com/user-attachments/files/27593185/OoT_72372_AKDSAIGODV.suns.song.patch.zip)
```
02 00 00 08 41 00 00 00
05 00 00 0c 4c 00 00 00
ff 00 00 08 00 00 00 00
05 00 00 0e 53 00 00 00
ff 00 00 0a 00 00 00 00
02 00 00 10 41 00 00 00
05 00 00 7f 4a 00 00 00
ff 00 00 01 00 00 00 00
ff 00 00 00 5a 00 00 00
00 00 00 00 00 00 00 00
```

(^ the second pause seems kinda off and I might have inserted it somehow but don't think so)

OoT_72372_LED20I0XVD
[OoT_72372_LED20I0XVD suns song patch.zip](https://github.com/user-attachments/files/27593193/OoT_72372_LED20I0XVD.suns.song.patch.zip)
```
02 00 00 0f 45 00 00 00
05 00 00 0c 43 00 00 00
ff 00 00 08 00 00 00 00
05 00 00 15 5e 00 00 00
02 00 00 18 52 00 00 00
05 00 00 7f 42 00 00 00
ff 00 00 01 00 00 00 00
ff 00 00 00 5a 00 00 00
00 00 00 00 00 00 00 00
```

**Comment by djevangelia** [comment]
I think I figured it out.

The "raw" song needs to be a length of 3n to bug
The final extra note needs to be 1 unit long (i.e. song length without it is 3n)

The issue is function `AudioOcarina_PlaybackSong`

```
u32 noteTimerStep;      // How much to decrease length of current note
u32 nextNoteTimerStep;  // How much to decrease length when starting next note

if (sPlaybackStaffPos == 0) {   // first note I assume?
        noteTimerStep = 3;
    } else {
        noteTimerStep = sOcarinaUpdateTaskStart - sOcarinaPlaybackTaskStart;    // Seems to always equal 3
    }

    if (noteTimerStep < sPlaybackNoteTimer) {
        sPlaybackNoteTimer -= noteTimerStep;    // If still lots of length left, just reduce
    } else {
        nextNoteTimerStep = noteTimerStep - sPlaybackNoteTimer;
        sPlaybackNoteTimer = 0; // !!!! Save how much was left of the note step to use for the next note
    }

    if (sPlaybackNoteTimer == 0) {  // Current note finished
        sPlaybackNoteTimer = sPlaybackSong[sPlaybackNotePos].length; // Load next note

        if (sPlaybackNotePos == 1) {
            sPlaybackNoteTimer++; // Increase the second note 1 unit for some reason???
        }

        if (sPlaybackNoteTimer == 0) {  // New length 0 means no further notes
            ...
            return;
        } else {
            sPlaybackNoteTimer -= nextNoteTimerStep; // !!!!! Otherwise, decrease the new note by the remaining note step
        }
```

If the final real note of the song ends up with 1 unit left before loading next note, `nextNoteTimerStep` becomes 2.
The final 1 unit note length  is loaded as the next `sPlaybackNoteTimer`. This results in 1 - 2 = underflow.

`noteTimerStep` always seems to be 3. To song length, the weird 1 unit addition for the second note must be added.
So, the final in-game song length needs to be 3n + 2 (including second note and last 1) to result in the final note always having 1 as final value giving rise to `nextNoteTimerStep = 2`.
From seed LED20I0XVD:
```
15 = 3n         -> OK
13 = 3n + 1     -> 2 left
8 = 3n + 2      -> OK (3n + 2 - 2)
21 = 3n         -> OK
24 = 3n         -> OK
127 = 3n + 1    -> 2 left
1 = 1           -> 1 - 2 = bad
```

This manipulated song, where every note length is 3n, is bugged:
```
[ 2, 00, 00,    0x27, 0x49, 00, 00, 00,
5, 00, 00,      0x27, 0x44, 00, 00, 00,
255, 00, 00,    0x0c, 00, 00, 00, 00,
5, 00, 00,      0x27, 0x42, 00, 00, 00,
2, 00, 00,      0x27, 0x5f, 00, 00, 00,
5, 00, 00,      0x27, 0x43, 00, 00, 00,
255, 00, 00,    0x1, 00, 00, 00, 00 ]
```
So fixing might be just setting a minimum note length for all notes including the final extra one. I don't know Python so I would rather stay away.
